knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go

```diff
@@ -69,38 +77,94 @@ type ReconcileSourceDeployment struct {
 	// that reads objects from the cache and writes to the apiserver
 	client client.Client
 	scheme *runtime.Scheme
+
+	testShouldUseClusterMonitoring          *bool
+	testShouldGenerateSourceServiceMonitors *bool
 }
 
 // Reconcile reads that state of the cluster for an eventing source deployment
 func (r *ReconcileSourceDeployment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling the source deployment, setting up a service/service monitor if required")
 	dep := &v1.Deployment{}
-	if err := r.client.Get(context.TODO(), request.NamespacedName, dep); err != nil {
-		// the deployment does not exist anymore do nothing
+	err := r.client.Get(context.TODO(), request.NamespacedName, dep)
+	inDeletion := false
+	if apierrors.IsNotFound(err) {
+		// The deployment does not exist anymore, deletions are shown as failing reads
+		log.Info("Source in deletion phase")
+		inDeletion = true
+	} else if err != nil {
+		// Do nothing if can't get the source for some other error
 		return reconcile.Result{}, nil
 	}
-	if err := monitoring.SetupMonitoringRequirements(r.client, dep); err != nil {
-		return reconcile.Result{}, err
-	}
-	if err := SetupSourceServiceMonitorResources(r.client, dep); err != nil {
+	eventing := &v1alpha1.KnativeEventing{}
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "knative-eventing", Name: "knative-eventing"}, eventing); err != nil {
 		return reconcile.Result{}, err
 	}
-	return reconcile.Result{}, nil
-}
-
-type skipDeletePredicate struct {
-	predicate.Funcs
+	shouldEnableMonitoring := okomon.ShouldEnableMonitoring(eventing.Spec.GetConfig())
+	return reconcile.Result{}, r.reconcileSourceMonitoring(r.client, dep, shouldEnableMonitoring, inDeletion)
 }
 
-func (skipDeletePredicate) Delete(e event.DeleteEvent) bool {
-	return false
+func (r *ReconcileSourceDeployment) reconcileSourceMonitoring(rclient client.Client, dep *v1.Deployment, shouldEnableMonitoring bool, inDeletion bool) error {
+	// Setup monitoring resources only if monitoring should be on and source deployment is not being deleted
```

Comment:
I'm wondering if we should extract some logic out of this method into other methods to explain the logic more and not rely on the detailed comments. Specially the the two branches of the big if statements? don't have a very strong opinion, just thinking out loud

Reply:
Let me think about it.